### PR TITLE
fix(localdebug): update run icon display status when the active text editor is changed

### PR DESCRIPTION
**Pull request by XiaofuHuang**
update run icon display status when the active text editor is changed

### CODE REVIEW

File: packages/vscode-extension/src/debug/runIconHandler.ts [hunk 1]
```diff
@@ -22,6 +22,7 @@ export async function selectAndDebug(args?: any[]): Promise<Result<null, FxError
 }
 
 export function registerRunIcon(): void {
+  ext.context.subscriptions.push(vscode.window.onDidChangeActiveTextEditor(enableRunIcon));
```
Comment (swatDong) [inline packages/vscode-extension/src/debug/runIconHandler.ts:25]: `isValidProject` is a bit heavy (contains I/O operation) so avoid to call it frequently.
How about only validate project when `onDidChangeWorkspaceFolders` and `onDidSaveTextDocument` of any *.fx/* file?

Comment (XiaofuHuang) [reply packages/vscode-extension/src/debug/runIconHandler.ts:25]: Good idea, All the condition has been changed.

Comment (swatDong) [reply packages/vscode-extension/src/debug/runIconHandler.ts:34]: `isValidProject` is required for correctly checking the whole project status, but just avoid calling it frequently.
Some options or you can add more:
1. consider `vscode.workspace.createFileSystemWatcher` to catch *.fx* changes, even out of vscode
2. set a timer to call `isValidProject` and save current status, and `onDidXxx` just refer the status flag

Comment (XiaofuHuang) [reply packages/vscode-extension/src/debug/runIconHandler.ts:34]: According to the [document](https://code.visualstudio.com/api/references/vscode-api), the FileSystemWatcher cannot track the folder change.
> Note that only files within the current workspace folders can be watched. Note that when watching for file changes such as '*/.js', notifications will not be sent when a parent folder is moved or deleted (this is a known limitation of the current implementation and may change in the future).

Comment (XiaofuHuang) [reply packages/vscode-extension/src/debug/runIconHandler.ts:35]: Add a cache. If the editor changed in 5s, the project status will not be checked.

File: packages/vscode-extension/src/debug/runIconHandler.ts [hunk 2]
```diff
@@ -22,12 +32,35 @@ export async function selectAndDebug(args?: any[]): Promise<Result<null, FxError
 }
 
 export function registerRunIcon(): void {
-  ext.context.subscriptions.push(vscode.workspace.onDidOpenTextDocument(enableRunIcon));
+  ext.context.subscriptions.push(vscode.window.onDidChangeActiveTextEditor(enableRunIcon));
   ext.context.subscriptions.push(vscode.workspace.onDidChangeWorkspaceFolders(enableRunIcon));
   enableRunIcon();
 }
 
+const projectStatusCache = new NodeCache({ stdTTL: 5 });
```
Comment (swatDong) [inline packages/vscode-extension/src/debug/runIconHandler.ts:40]: 5 seconds? a bit small..

**Review by swatDong — APPROVED** [review]

File: packages/vscode-extension/src/debug/runIconHandler.ts [hunk 3]
```diff
@@ -1,14 +1,25 @@
...
+import * as path from "path";
+import { ExtTelemetry } from "../telemetry/extTelemetry";
+import { TelemetryEvent, TelemetryProperty } from "../telemetry/extTelemetryEvents";
 
-export async function selectAndDebug(args?: any[]): Promise<Result<null, FxError>> {
+export async function selectAndDebug(): Promise<Result<null, FxError>> {
   if (ext.workspaceUri && isValidProject(ext.workspaceUri.fsPath)) {
-    vscode.commands.executeCommand("workbench.view.debug");
-    vscode.commands.executeCommand("workbench.action.debug.selectandstart");
+    await vscode.commands.executeCommand("workbench.view.debug");
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/debug/runIconHandler.ts:21]: Do we need to await here, what is the point?

File: packages/vscode-extension/src/debug/runIconHandler.ts [hunk 4]
```diff
@@ -22,12 +33,51 @@ export async function selectAndDebug(args?: any[]): Promise<Result<null, FxError
 }
 
 export function registerRunIcon(): void {
-  ext.context.subscriptions.push(vscode.workspace.onDidOpenTextDocument(enableRunIcon));
-  ext.context.subscriptions.push(vscode.workspace.onDidChangeWorkspaceFolders(enableRunIcon));
-  enableRunIcon();
+  ext.context.subscriptions.push(
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/debug/runIconHandler.ts:36]: Is it better to just set the status of whether project is valid teamsfx project and don't care about whether active text editor changed. Do you use onDidChangeActiveTextEditor() here for situations when user checkout code or something else?

Comment (XiaofuHuang) [reply packages/vscode-extension/src/debug/runIconHandler.ts:36]: Yes, the user may change the project, and the run icon status will be changed.
When the project is open, the project status will be force checked

File: packages/vscode-extension/src/debug/runIconHandler.ts (hunk 3 above)
Comment (XiaofuHuang) [reply packages/vscode-extension/src/debug/runIconHandler.ts:21]: Here we execute two commands, if we do not use await, the command `workbench.view.debug` may be execute later than `workbench.action.debug.selectandstart`, which may case the options to be closed.

**Review by HuihuiWu-Microsoft — APPROVED** [review]

**Pull request closed without merging**